bazel_7: init at 7.0.0

layus opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Introduce bazel 7

What works ?

- building a functional bazel binary and most tests for
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [X] aarch64-darwin
- [X] Extracting MODULE.bazel.lock into a repository_cache / bindir nix derivation :tada: 

What does not work ?

  - [x] Trimming bazel external repos required for tests
  - [x] Protobuf tests (full rewrite I think)
  - [x] MODULE.lock generation is broken, a GSON issue I guess.
  - [x] Java toolchain is hardcoded using aggressive system bazelrc.
  - [x] cleanup of useless files
  - [x] Comment every piece for future me/us
  - [x] (re)Implement updater
  - [ ] Split the bootstrap steps into different derivations (I have loost too much time failing in the last ten seconds after 10 minutes of compiling)

More details in a follow-up comment, but here is a PR with two weeks of work, to avoid duplicating work.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [X] aarch64-darwin
- [X] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

layus committed:
bazel_7: init at 7.0.0-pre.20230917.3

layus committed:
bazel_7: start by copying bazel_6

layus committed:
bazel_7: Simplify callBazelTest

layus committed:
bazel_7: restrict USER hack to darwin

layus committed:
bazel_7: split tests for clarity

layus committed:
bazel_7: refactor and comment

layus committed:
bazel_7: More refactors

layus committed:
Get bazel building on all major platforms, with tests

layus committed:
bazel_7: Filter bzlmod dependencies by name predicate

layus committed:
bazel_7: Separate test deps (without binaries)

layus committed:
bazel_7: Cleanup patchPhase

layus committed:
Rework repository_cache, lockfile generation, and protobuf tests as a whole

layus committed:
Add 7.0.0rc2, but some patches need an update

layus committed:
Bump to 7.0.0rc2

layus committed:
bazel_7: cleanup changes made to common tests

layus committed:
bazel_7: cleanup changes made to common patches

layus committed:
bazel_7: cleanup protobuf tests

layus committed:
bazel_7: cleanup unused files

layus committed:
bazel_7: rename protobuf test lockfile

layus committed:
bazel_7: rename protobuf test lockfile, part 2

layus commented:
This is now ready for review. Any insight welcome.

I did test it with 

```
nix-build -E'(import ./. { system = "'{{x86_64,aarch64}-darwin,x86_64-linux}'"; })'.bazel_{4,5,6,7}.tests
```

And while bazel_6 fails to build on darwin, I consider this pretty unrelated, as it fails to find codesign, so probably a sandbox / patching issue.

Tests do pass on linux for all bazel versions, and on darwin for bazel_7.

layus committed:
bazel_7: darwin (sandbox) fixes

uri-canva reviewed (changes requested):
Awesome work! I'm afraid the nix functions are a bit beyond my skills, I can understand them and I believe they're correct but I don't know if there's ways to make them better / more idiomatic. Hopefully someone else can take a look as well, but either way I don't think there's any blockers there.

boltzmannrain commented:
`aarch64-linux` compiles too here :+1:

layus commented:
I know I am long overdue here, but there are still some pending issues, and we are very close to bazel 7.0.0 official release. I will try to release that directly.

layus committed:
bazel_7: 7.0.0rc2 -> 7.0.0rc7

boltzmannrain commented:
If merged with latest master it starts to fail on darwin like so
`x86_64-darwin`:
```
...
ld: warning: dylib (/nix/store/x4s0wi6d2z38wa0jb4yf769sl1c0c50z-libcxx-16.0.6/lib/libc++.dylib) was built for newer macOS version (10.12) than being linked (10.11)
Undefined symbols for architecture x86_64:
  "vtable for std::out_of_range", referenced from:
      std::out_of_range::out_of_range[abi:v160006](char const*) in libijar_cc_binary_main.lo(classfile.o)
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "___cxa_end_catch", referenced from:
      std::__1::basic_stringbuf<char, std::__1::char_traits<char>, std::__1::allocator<char>>::overflow(int) in libijar_cc_binary_main.lo(classfile.o)
      std::__1::basic_ostream<char, std::__1::char_traits<char>>& std::__1::__put_character_sequence[abi:v160006]<char, std::__1::char_traits<char>>(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, char const*, unsigned long) in libijar_cc_binary_main.lo(classfile.o)
...
```
`aarch64-darwin`:
```
...
Undefined symbols for architecture arm64:
  "vtable for std::out_of_range", referenced from:
      std::out_of_range::out_of_range[abi:v160006](char const*) in libijar_cc_binary_main.lo(classfile.o)
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "___cxa_end_catch", referenced from:
      std::__1::basic_stringbuf<char, std::__1::char_traits<char>, std::__1::allocator<char>>::overflow(int) in libijar_cc_binary_main.lo(classfile.o)
      std::__1::basic_ostream<char, std::__1::char_traits<char>>& std::__1::__put_character_sequence[abi:v160006]<char, std::__1::char_traits<char>>(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, char const*, unsigned long) in libijar_cc_binary_main.lo(classfile.o)
  "typeinfo for std::bad_array_new_length", referenced from:
      std::__throw_bad_array_new_length[abi:v160006]() in libijar_cc_binary_main.lo(classfile.o)
...
```

layus commented:
Thanks for the report, I will rebase and try to fix. Any idea what the fix could be though ? I have access to a Darwin builder, but not much knowledge.

Le 9 décembre 2023 12:25:29 GMT+01:00, Dmitry Ivankov ***@***.***> a écrit :
>If merged with latest master it starts to fail on darwin like so
>`x86_64-darwin`:
>```
>...
>ld: warning: dylib (/nix/store/x4s0wi6d2z38wa0jb4yf769sl1c0c50z-libcxx-16.0.6/lib/libc++.dylib) was built for newer macOS version (10.12) than being linked (10.11)
>Undefined symbols for architecture x86_64:
>  "vtable for std::out_of_range", referenced from:
>      std::out_of_range::out_of_range[abi:v160006](char const*) in libijar_cc_binary_main.lo(classfile.o)
>  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
>  "___cxa_end_catch", referenced from:
>      std::__1::basic_stringbuf<char, std::__1::char_traits<char>, std::__1::allocator<char>>::overflow(int) in libijar_cc_binary_main.lo(classfile.o)
>      std::__1::basic_ostream<char, std::__1::char_traits<char>>& std::__1::__put_character_sequence[abi:v160006]<char, std::__1::char_traits<char>>(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, char const*, unsigned long) in libijar_cc_binary_main.lo(classfile.o)
>...
>```
>`aarch64-darwin`:
>```
>...
>Undefined symbols for architecture arm64:
>  "vtable for std::out_of_range", referenced from:
>      std::out_of_range::out_of_range[abi:v160006](char const*) in libijar_cc_binary_main.lo(classfile.o)
>  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
>  "___cxa_end_catch", referenced from:
>      std::__1::basic_stringbuf<char, std::__1::char_traits<char>, std::__1::allocator<char>>::overflow(int) in libijar_cc_binary_main.lo(classfile.o)
>      std::__1::basic_ostream<char, std::__1::char_traits<char>>& std::__1::__put_character_sequence[abi:v160006]<char, std::__1::char_traits<char>>(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, char const*, unsigned long) in libijar_cc_binary_main.lo(classfile.o)
>  "typeinfo for std::bad_array_new_length", referenced from:
>      std::__throw_bad_array_new_length[abi:v160006]() in libijar_cc_binary_main.lo(classfile.o)
>...
>```
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/NixOS/nixpkgs/pull/262152#issuecomment-1848380021
>You are receiving this because you were assigned.
>
>Message ID: ***@***.***>

aaronmondal commented:
7.0.0 was released: https://github.com/bazelbuild/bazel/releases

boltzmannrain commented:
Looks like 7.x darwin build breaks after https://github.com/NixOS/nixpkgs/pull/263535

boltzmannrain commented:
```
 (cd /private/tmp/nix-build-bazel-7.0.0.drv-3/bazel_eVl8TlHS/out/execroot/_main && \
  exec env - \
    PATH=/nix/store/ai4hd8f1xhr0rfjdr17bxx1rwi42sx97-python3-3.11.6/bin
/nix/store/bg77k830j2z1xjslll7d31y55d3hnmip-unzip-6.0/bin
/nix/store/79yyc6yisbxkcpgavm2c756ysy5r3jbm-which-2.21/bin
/nix/store/pwhi3qhn5fi7m64hxyn2wkgy3b4l91zi-zip-3.0/bin
/nix/store/xd8i1lilimkhn4zcxyd3icddzb7r088z-cctools-port-973.0.1/bin
/nix/store/3fpl3hs6ii820yxp0rbafx5phnmj8ng7-clang-wrapper-16.0.6/bin
/nix/store/a5v30qll5i02vr9y97bk1rdx3mm6kvlm-clang-16.0.6/bin
/nix/store/h0fgpyxfav7ybnw5sdg8jpyv9f95x4w0-coreutils-9.3/bin
/nix/store/ydcx9rj4gf3dxm1kc0lx7l5d5vrjdwp2-cctools-binutils-darwin-wrapper-16.0.6-973.0.1/bin
/nix/store/vpnp0c420cjmx82g05jq3nm5skdrblvk-cctools-binutils-darwin-16.0.6-973.0.1/bin
/nix/store/k5xqpf63xm8m0v8m0bsb34w4w4l7yids-zulu17.44.53-ca-jdk-17.0.8.1/bin
/nix/store/x1xxsh1gp6y389hyl40a0i74dkxiprl7-bash-5.2-p15/bin
/nix/store/94j70ziq1zp2mj0fs2fzi49j7q3mdhba-diffutils-3.10/bin
/nix/store/ckbx2qla8zxic5k2d549qdhkbwrpa4q6-file-5.45/bin
/nix/store/zr6klxfjzpdr2674ly1f4fix7ig57mjr-findutils-4.9.0/bin
/nix/store/blggp459hvq6swvr8nlxblkdn7ayw8y7-gawk-5.2.2/bin
/nix/store/98v0v3bwzgkcwsz4anjip62d2k3gz352-gnugrep-3.11/bin
/nix/store/zapwwg8vnc47j7ciyrd7hb7rhwfrrs5c-patch-2.7.6/bin
/nix/store/nddzbf5pifm77kksh25xwiy2g9kqqm9f-gnused-4.9/bin
/nix/store/g9z0hg6snhg5834p14yw4bx6ilcffr8b-gnutar-1.35/bin
/nix/store/6g3nq5hply5p2v3xzlg0i9qqgzr30w1d-gzip-1.13/bin
/nix/store/h0fgpyxfav7ybnw5sdg8jpyv9f95x4w0-coreutils-9.3/bin
/nix/store/zr6klxfjzpdr2674ly1f4fix7ig57mjr-findutils-4.9.0/bin
/nix/store/94j70ziq1zp2mj0fs2fzi49j7q3mdhba-diffutils-3.10/bin
/nix/store/nddzbf5pifm77kksh25xwiy2g9kqqm9f-gnused-4.9/bin
/nix/store/98v0v3bwzgkcwsz4anjip62d2k3gz352-gnugrep-3.11/bin
/nix/store/blggp459hvq6swvr8nlxblkdn7ayw8y7-gawk-5.2.2/bin
/nix/store/g9z0hg6snhg5834p14yw4bx6ilcffr8b-gnutar-1.35/bin
/nix/store/6g3nq5hply5p2v3xzlg0i9qqgzr30w1d-gzip-1.13/bin
/nix/store/dq6dsprg57r5wqlc4yzcnkbfxdhfcd3f-bzip2-1.0.8-bin/bin
/nix/store/6kxdgmbgwdq4y42r9qhgdky312v5d7x0-gnumake-4.4.1/bin
/nix/store/x1xxsh1gp6y389hyl40a0i74dkxiprl7-bash-5.2-p15/bin
/nix/store/j34fimsig5s1zsf2f5dkw9a8sv5zpskg-patch-2.7.6/bin
/nix/store/96bznf6bjdghshgj3a2hm16gk3nngbcx-xz-5.4.4-bin/bin
/nix/store/ckbx2qla8zxic5k2d549qdhkbwrpa4q6-file-5.45/bin \
    PWD=/proc/self/cwd \
    ZERO_AR_DATE=1 \
  external/bazel_tools~cc_configure_extension~local_config_cc/cc_wrapper.sh @bazel-out/darwin_arm64-opt-exec-ST-fad1763555eb/bin/external/rules_java~7.1.0~toolchains~remote_java_tools/ijar_cc_binary-2.params)
# Configuration: ab192274ab21bdc5bb8e345e4cc8bd36536328aa117b8df888873da443d114d6
# Execution platform: //:default_host_platform
Undefined symbols for architecture arm64:
  "vtable for std::out_of_range", referenced from:
      std::out_of_range::out_of_range[abi:v160006](char const*) in libijar_cc_binary_main.lo(classfile.o)
...  
```
It looks like there's some instability around libcxx and apple sdk version that is going to be addressed in master (after 23.11 release, so ~now). But not exactly sure what goes wrong here

dmivankov commented:
Seems to be https://github.com/NixOS/nixpkgs/issues/166205
And might be fixable with something like https://github.com/NixOS/nixpkgs/pull/268829/commits/51451ac4069e18b5813b38d57cb68cd72b7ec51e

boltzmannrain committed:
bazel_7: fix darwin compilation with CLang 16 & recent nixpkgs/master

Workaround for #166205 similar to
https://github.com/NixOS/nixpkgs/commit/51451ac4069e18b5813b38d57cb68cd72b7ec51e

And warning fix similar to `bazel_6` fix for
```
external/upb~0.0.0-20220923-a547704/upb/mini_table.c:634:14: error: defining a type within '__builtin_offsetof' is a Clang extension [-Werror,-Wgnu-offsetof-extensions]
  UPB_ASSERT(UPB_ALIGN_OF(upb_StringView) ==
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

boltzmannrain commented:
https://github.com/layus/nixpkgs/pull/12 fixes compilation for me with recent nixpkgs/master on aarch64-darwin

layus committed:
bazel_7: 7.0.0rc7 -> 7.0.0

layus commented on pkgs/development/tools/build-managers/bazel/bazel_7/default.nix:
```diff
@@ -0,0 +1,603 @@
+    #     '';
+    #
+    # I do not know yet how to allow IOPMAssertion{CreateWithName,Release}
+    ./darwin_sleep.patch
```
It would make it impossible to test bazel_7 _inside_ a nix derivation. (Or am I mixing things up ?)

layus commented on pkgs/development/tools/build-managers/bazel/bazel_7/default.nix:
```diff
@@ -0,0 +1,603 @@
+  bazelRC = writeTextFile {
+    name = "bazel-rc";
+    text = ''
+      startup --server_javabase=${buildJdk}
```
Not quite sure yet. I think we have issues when buildJdk != runJdk. Generated lockfiles are always empty if some versions do not align. Have to investigate still.

layus replied:
Fixed, and tested that it works.
@avdv , you might be interested in [`2a4700c` (#262152)](https://github.com/NixOS/nixpkgs/pull/262152/commits/2a4700c48be3c6b58e06fc31a78924fe643a6bae) (not a proper fix, just a test based on your dicovery

layus committed:
bazel_7: address most review comments

layus committed:
bazel_7: restore xcode_locator patch from no-arc patch

layus committed:
bazel_7: restore xcode_locator no-arc fix, still needed

layus committed:
bazel_7: re-enable arc in xcode_locator command

layus committed:
bazel_7: fix cpp-test with bzlmod, for linux

layus committed:
bazel_7: fix protobuf test on darwin

FollieHiyuki commented:
Does this `bazel_7` package also suffers from https://github.com/NixOS/nixpkgs/issues/273500? I've not tested it yet. Just wondering.

layus commented:
Oddly enough no, but I observed it a few times. I think it boils down to getting the right Java version to compile Bazel, with the right Java language version flags. But I am not sure.

Le 14 décembre 2023 16:11:16 GMT+01:00, Hoang Nguyen ***@***.***> a écrit :
>Does this `bazel_7` package also suffers from https://github.com/NixOS/nixpkgs/issues/273500? I've not tested it yet. Just wondering.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/NixOS/nixpkgs/pull/262152#issuecomment-1856028534
>You are receiving this because you were assigned.
>
>Message ID: ***@***.***>

malt3 commented:
I think this doesn't have the `bashWithDefaultShellUtils` as a shell instead of `${bash}/bin/bash` that was recently added to bazel_6:

https://github.com/NixOS/nixpkgs/blob/459e800b0475d6804808798f67a5f58e3294a6e4/pkgs/development/tools/build-managers/bazel/bazel_6/default.nix#L452-L454

Does this introduce a regression for #265040? If so: Is this intended?

layus committed:
bazel_7: backport bazel_6 bash fixes for remote execution

layus committed:
bazel_7: Unse runJDK for runtime

layus committed:
fixup! bazel_7: backport bazel_6 bash fixes for remote execution

layus committed:
bazel_7: Add regression test for empty lockfiles

layus commented:
@FollieHiyuki I have added a test for https://github.com/NixOS/nixpkgs/issues/265040 to prevent regressions. At least the binary we produce will not have the issue, while the origin is still unknown.

@malt3 "Backported" remote build fixes from bazel_6, and removed the corresponding patch. Would you mind testing it for remote execution. I never encountered that issue with remote execution myself.

layus committed:
fixup! bazel_7: backport bazel_6 bash fixes for remote execution

malt3 reviewed (approved):
> Would you mind testing it for remote execution.

This backport looks good to me and appears to work fine. Thank you!

name-snrl commented:
@layus Hello, when I tried to use `bazel-repository-cache.nix` without passing the `requiredDepNamePredicate` argument, it threw an error that the value is a set, while a string was expected. The problem is in `_main~bazel_bazel_build_deps~workspace_repo_cache`, because sha256 here is an array with the pair `archive -> hash`, instead of str.

I don't know how to fix this in a better way or if it should be done. But I figured it should be reported.

layus committed:
bazel_7: Cleanup tests a bit

layus committed:
bazel_7: improve lockfile parsing

layus committed:
bazel_7: fix remaining tests

layus commented:
@name-snrl I have fixed that use case, because I needed the files it bundled. I cannot however support all repository_rule attributes, so the script should be made more robust to failure. I have improved that a bit.

layus committed:
bazel_7: Check-in the main lockfile

layus commented:
Okay, so now all tests pass :tada: 

As long as ofborg is happy, I am going to merge (or let anyone else merge).

layus committed:
fixup! bazel_7: Check-in the main lockfile

timothyklim commented:
Is there any reason why remote_java_tools_linux not patcheelf'ed?
```
Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
src/main/tools/linux-sandbox-pid1.cc:530: "execvp(external/remote_java_tools_linux/java_tools/ijar/ijar, 0x1a6db30)": No such file or directory
```

layus commented:
It is not because patching Bazel to apply patchelf on downloaded tarballs is an heavy change, and probably unexpected by users.

There is however a fix for that issue which relies on using a local Java toolchain and recompiling ijar from sources. The core parameters are in the hardcoded bazelrc file near the top of the nix file.

Ensure you have JAVA_HOME set, and use toolchain resolution debug flag to see why the non_prebuilt toolchain is not used

Le 3 janvier 2024 03:54:07 GMT+01:00, timothy ***@***.***> a écrit :
>Is there any reason why remote_java_tools_linux not patcheelf'ed?
>```
>Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
>src/main/tools/linux-sandbox-pid1.cc:530: "execvp(external/remote_java_tools_linux/java_tools/ijar/ijar, 0x1a6db30)": No such file or directory
>```
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/NixOS/nixpkgs/pull/262152#issuecomment-1874790433
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>

Strum355 commented:
Is bazel7 in nixpkgs missing the [action_env patch](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/build-managers/bazel/bazel_6/actions_path.patch) that exists in bazel6? I seem to be hitting the case that was solved in bazel 6 by https://github.com/NixOS/nixpkgs/pull/146389 after upgrading to bazel7, where `build --action_env=PATH=/nix/store/...coreutils:/nix/store/...stuff` isnt being propagated to `ctx.action.run`

rickvanprim commented:
@Strum355 I hit a few cases last night during my upgrade to Bazel 7 where things in `actions.run()` is missing `coreutils` and setting `use_default_shell_env=True` fixes it but I'm not sure if that's correct.  Without it, it looks like the command is being run with zero environment (whereas in Bazel 6, the patched `PATH` was still there).  I suspect needing `use_default_shell_env=True` is wrong though because I'm not sure how non-Nix usages would work either without some base amount of `PATH` given how many things assume the presence of `coreutils`.

layus commented:
I had the same problem with my deployment. And I just re-added the patch (locally) for now. What bugs me is that the environment ends up empty, but I see no path yet in the code where it can happen.

Re-adding the patch is the quickest fix.

My best guess is that the patch made us use run() in an incorrect way, assuming that coreutils is present when it is not by default. Bazel add PATH to the action digest, so for tools that need no path, it's better to have none. The patch fooled us into thinking that run() can be used that way. Not too sure yet, I need to investigate using a vanilla bazel.

Oh, and not gonna work on that in the next few days sadly, but probably end of next week.

Strum355 commented:
> Is bazel7 in nixpkgs missing the [action_env patch](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/build-managers/bazel/bazel_6/actions_path.patch?rgh-link-date=2024-01-05T13%3A40%3A35Z) that exists in bazel6? I seem to be hitting the case that was solved in bazel 6 by #146389 after upgrading to bazel7, where `build --action_env=PATH=/nix/store/...coreutils:/nix/store/...stuff` isnt being propagated to `ctx.action.run`

Correction on this, it appears that what I thought was a case of this patch causing the passing through action_env wasnt actually the case, but it was masking the fact that action_env wasnt being passed through by hard-coding just enough to make some rules that depend on e.g. coreutils run successfully. A lot of convolution around action_env and where/what passes through $PATH, but the patch is just "good enough" for me to keep working. Ideally everything would pass down action_env, as it would mean I could avoid rebuilding bazel by having to to set a different `defaultShellPath` :/

boltzmannrain commented:
With `--repo_env` and `--action_env` it's possible to set `PATH` and even `LD_LIBRARY_PATH` but some rules override that and pretend to be in control while in reality they rely on implicit non-hermetic dependencies in FHS locations.

Another rabbit hole with this is it's possible to undo nixos glibc&bazel patches, tweak bazel build once, and then run builds in specially-crafted buildFHSEnv. So rules that expect to work with empty env/path will fallback to fhs through glibc (hence need to unpatch it, nixos disables fhs fallback) and find whatever is put into fhs (and needs to be in fhs as this is where glibc would look on empty env). To make it bazel cache friendly probably still makes sense to add `--repo_env` and `--action_env` to `bazelrc`
